docs/compilers/CSharp/Runtime Async Design.md

```diff
@@ -0,0 +1,456 @@
+# Runtime Async Design
+
+See also the ECMA-335 specification change for this feature: https://github.com/dotnet/runtime/pull/104063, https://github.com/dotnet/runtime/blob/main/docs/design/specs/runtime-async.md (when the PR completes). https://github.com/dotnet/runtime/issues/109632 tracks open issues for the feature in the runtime.
+
+This document goes over the general design of how Roslyn works with the Runtime Async feature to produce IL. In general, we try to avoid exposing this feature at the user level; initial binding is almost entirely
+unaffected by runtime async. Exposed symbols do not give direct information about whether they were compiled with runtime async, and indeed the compiler has no idea whether a method from a referenced assembly is
+compiled with runtime async or not.
+
+## Supporting runtime apis
+
+We use the following runtime flag to drive whether feature can be used. This flag must be defined in the same assembly that defines `object`, and the assembly cannot reference any other assemblies. In terms of
+CoreFX, this means it must be defined in the `System.Runtime` reference assembly.
+
+TODO: Determine whether just the presence of this flag will cause the compiler to generate in runtime async mode.
+
+```cs
+namespace System.Runtime.CompilerServices;
+
+public static class RuntimeFeature
+{
+    public const string Async = nameof(Async);
+}
+```
+
+We use the following helper APIs to indicate suspension points to the runtime, in addition to the runtime async call syntax:
+
+```cs
+namespace System.Runtime.CompilerServices;
+
+// These methods are used to await things that cannot use runtime async signature form
+// TODO: Clarify which of these should be preferred? Should we always emit the `Unsafe` version when awaiting something that implements `ICriticalNotifyCompletion`?
+namespace System.Runtime.CompilerServices;
+
+public static class RuntimeHelpers
+{
+    [RuntimeAsyncMethod]
+    public static Task AwaitAwaiterFromRuntimeAsync<TAwaiter>(TAwaiter awaiter) where TAwaiter : INotifyCompletion;
+    [RuntimeAsyncMethod]
+    public static Task UnsafeAwaitAwaiterFromRuntimeAsync<TAwaiter>(TAwaiter awaiter) where TAwaiter : ICriticalNotifyCompletion;
+}
+```
+
+Additionally, we use the following helper attributes to indicate information to the runtime. If these attributes are not present in the reference assemblies, we will generate them; the runtime matches by full
+name, not by type identity, so we do not need to care about using the "canonical" versions.
+
+```cs
+namespace System.Runtime.CompilerServices;
+
+// Used to tell the runtime to generate the async state machinery for this method
+[AttributeUsage(AttributeTargets.Method)]
+public class RuntimeAsyncMethodAttribute() : Attribute();
```

Comment:
It would be good to have an example of C# code and corresponding emitted signature with this attribute.
I think that signature should also be emitted with the Experimental attribute in .NET 10.

Reply:
> It would be good to have an example of C# code and corresponding emitted signature with this attribute.

Not sure what you're asking for here. The method signature is unchanged, the attribute is just added.

> I think that signature should also be emitted with the Experimental attribute in .NET 10.

Presumably, we still need to decide exactly what the semantics will be here.

> Are we planning an update (or ask for an update) to ILVerify?

I believe that was discussed, though I don't remember who was taking it on. @agocke, do you know?